### PR TITLE
Added efficient binary search on loop range search

### PR DESCRIPTION
**Pull request by tanzircoderrect**
The current version uses a linear search on the loop ranges. These loop ranges are disjoint and created in the order of when their ends are encountered. As a result, these ranges are already sorted, which supports binary search. Therefore, all the linear searches are replaced by binary search function (i.e., lower_bound). A comparator RegionLessThan was added in the .h that compares two ranges based on the "end" value. This makes lower_bound() put an eventID in the right range if it already exists or in a higher range, which means eventID does not belong.

### CODE REVIEW

File: src/Analysis/OpenMPAnalysis.h [hunk 1]
```diff
@@ -26,6 +26,13 @@ struct Region {
 
   inline bool contains(EventID e) const { return end >= e && e >= start; }
 };
+struct RegionLessThan{
+    inline bool operator() (const Region& struct1, const Region& struct2)
+    {
+        return (struct1.end < struct2.end);
+    }
+};
+
```
Comment (BradSwain) [inline src/Analysis/OpenMPAnalysis.h:35]: Can this be made into a function and put inside of an anonymous namespace in OpenMPAnalysis.cpp?

```cpp
bool regionEndLessThan(const Region& region1, const Region& region2) { return region1.end < region2.end; }
```

File: src/Analysis/OpenMPAnalysis.cpp [hunk 2]
```diff
@@ -159,10 +158,11 @@ const std::vector<OpenMPAnalysis::LoopRegion>& OpenMPAnalysis::getOmpForLoops(co
 bool OpenMPAnalysis::inParallelFor(const race::MemAccessEvent* event) {
   auto loopRegions = getOmpForLoops(event->getThread());
   auto const eid = event->getID();
-  for (auto const& region : loopRegions) {
-    if (region.contains(eid)) return true;
-    // Break early if we pass the eid without finding matching region
-    if (region.end > eid) return false;
+  Region temp (eid, eid);
+  auto it = lower_bound (loopRegions.begin(), loopRegions.end(), temp, RegionLessThan());
```
Comment (BradSwain) [inline src/Analysis/OpenMPAnalysis.cpp:162]: Can the `temp` passed to `lower_bound` here be replaced with `Region(eid, eid)` directly to match the other calls to lower_bound?

**Review by BradSwain — CHANGES_REQUESTED** [review]
Looks great! Just a couple of small changes requested and then it looks good to me.

File: src/Analysis/OpenMPAnalysis.cpp (hunk 2 above)
Comment (tanzircoderrect) [reply src/Analysis/OpenMPAnalysis.cpp]: Done

**Review by BradSwain — APPROVED** [review]

**Pull request merged**